doc: doc example to use ajv-errors (#2254)
fastify/fastify: docs/Validation-and-Serialization.md

})
```

If you want custom error response in schema without headaches and quickly, you can take a look at [`ajv-errors`](https://github.com/epoberezkin/ajv-errors).

> Below is an example showing how to add custom error messages for each property of a schema by supplying a custom AJV instance. Inline comments describe how to can configure the schema to show a different error message for each case:

Reviewer: ```suggestion
Below is an example showing how to add custom error messages for each property of a schema by supplying a custom AJV instance. Inline comments describe how to configure the schema to show a different error message for each case:
```

Oops. Missed one.

})
```

> If you want custom error response in schema without headaches and quickly, you can take a look at [here](https://github.com/epoberezkin/ajv-errors).
> 
> Below is an example to add custom error messages for each property. Inline comments describe how you can configure your schema to show different error message for each case

Reviewer: ```suggestion
If you want custom error response in schema without headaches and quickly, you can take a look at [`ajv-errors`](https://github.com/epoberezkin/ajv-errors).

Below is an example showing how to add custom error messages for each property of a schema by supplying a custom AJV instance. Inline comments describe how to can configure the schema to show a different error message for each case:
```

```js
const Ajv = require('ajv')

> const AjvErrors = require('ajv-errors')

Reviewer: This is non-idiomatic. The module exports a regular function, not a "constructor function".
```suggestion
const ajvErrors = require('ajv-errors')
```

> ...

Reviewer: ```suggestion
```

const schema = {
  body: {
    type: 'object',
    properties: {
      name: {
        type: 'string',
        errorMessage: {
          type: 'Bad name'
        }
      },
      age: {
        type: 'number',
        errorMessage: {
          type: 'Bad age', // specify custom message for
          min: 'Too young' // all constraints except required
        }
      }
    },
    required: ['name', 'age'],
    errorMessage: {
      required: {
        name: 'Why no name!', // specify error message for when the
        age: 'Why no age!' // property is missing from input
      }
    }
  }
}

> ...

Reviewer: ```suggestion
```

const ajv = new Ajv({ allErrors: true, jsonPointers: true })

> // enhance the ajv instance
> AjvErrors(ajv)

Reviewer: ```suggestion
// extend the AJV instance with custom errors support
ajvErrors(ajv)
```

      name: {
        type: 'string',
        errorMessage: {
          type: 'Bad name'
        }
      },
      age: {
        type: 'number',
        errorMessage: {
          type: 'Bad age', // specify custom message for
          min: 'Too young' // all constraints except required
        }
      }
    },
    required: ['name', 'age'],
    errorMessage: {
      required: {
        name: 'Why no name!', // specify error message for when the
        age: 'Why no age!' // property is missing from input
      }
    }
  }
}

fastify.post('/', { schema, }, (request, reply) => {
  reply.send({
    hello: 'world'
  })
})
```

If you want to return localized error messages, take a look at [ajv-i18n](https://github.com/epoberezkin/ajv-i18n)

```js
const localize = require('ajv-i18n')

const fastify = Fastify({
  ajv: {
    customOptions: { allErrors: true }
  }
})

const schema = {
  body: {
    type: 'object',
    properties: {
      name: {
        type: 'string',
      },
      age: {
        type: 'number',
      }
    },
    required: ['name', 'age'],
  }
}

fastify.setErrorHandler(function (error, request, reply) {
  if (error.validation) {
    localize.ru(error.validation)

>     reply.status(400).send(error.validation)
>   }

Reviewer 2: ```suggestion
    reply.status(400).send(error.validation)
    return
  }
  reply.send(error)
```

})
```

If you want custom error response in schema without headaches and quickly, you can take a look at [`ajv-errors`](https://github.com/epoberezkin/ajv-errors).

> Below is an example showing how to add custom error messages for each property of a schema by supplying a custom AJV instance. Inline comments describe how to configure the schema to show a different error message for each case:

Reviewer 2: I would add this example [link](https://github.com/fastify/example/tree/master/validation-messages#example-of-custom-validation-errors-using-ajv-and-fastify-v2) as "quick play".

Author: You mean remove the example snippet from here and only have the link to example?

Reviewer 2: No, it is ok the example here 👍

I would like to give to that example more visibility

```js
const Ajv = require('ajv')
const ajvErrors = require('ajv-errors')

const schema = {
  body: {
    type: 'object',
    properties: {
      name: {
        type: 'string',
        errorMessage: {
          type: 'Bad name'
        }
      },
      age: {
        type: 'number',
        errorMessage: {
          type: 'Bad age', // specify custom message for
          min: 'Too young' // all constraints except required
        }
      }
    },
    required: ['name', 'age'],
    errorMessage: {
      required: {
        name: 'Why no name!', // specify error message for when the
        age: 'Why no age!' // property is missing from input
      }
    }
  }
}

> const ajv = new Ajv({ allErrors: true, jsonPointers: true })

Reviewer 2: I would remove the ajv custom instance and bet on these settings instead: https://github.com/fastify/fastify/blob/2.x/docs/Server.md#ajv

```
const fastify = require('fastify')({
  ajv: {
    customOptions: { allErrors: true, jsonPointers: true },
    plugins: [
      require('ajv-errors')
    ]
  }
})
```

// extend the AJV instance with custom errors support
ajvErrors(ajv)

> fastify.setSchemaCompiler(function (schema) {

Reviewer 2: thanks to the `ajv.customOptions` this is not needed anymore

  return ajv.compile(schema)
})
```

If you want to return localized error messages, take a look at [ajv-i18n](https://github.com/epoberezkin/ajv-i18n)

```js
const Ajv = require('ajv')
const localize = require('ajv-i18n')

const schema = {
  body: {
    type: 'object',
    properties: {
      name: {
        type: 'string',
      },
      age: {
        type: 'number',
      }
    },
    required: ['name', 'age'],
  }
}

const ajv = new Ajv({ allErrors: true, jsonPointers: true })

fastify.setSchemaCompiler(function (schema) {
  const validate =  ajv.compile(schema)

  function validator (data) {
    const result = validate(data)

    if (!result) {
      localize.ru(validate.errors);

>       ajv.errorsText(validate.errors, { separator: '\n' })

Reviewer 2: I would manage this in a different manner after creating the `fastify` instance as above.
Less code to maintain for the user and it could become a plugin to reuse in other application

```
fastify.setErrorHandler(function (error, request, reply) {
  if (error.validation) {
     localize.ru(error.validation)
     fastify.schemaCompiler.errorsText(error.validation, { separator: '\n' })
     reply.status(400).send(error.validation)
  }
})
```

Author: `errorsText` is not on `fastify.schemaCompiler`. It's on `ajv`'s instance. Whereas `fastify.schemaCompiler` is a compiler's instance, which is a function not an object.

So I've kept the custom ajv instance and using it in `errorHandler`. Let me know if you have any other way of accessing ajv's instance.

Reviewer 2: You are right, but we can just remove the `fastify.schemaCompiler.errorsText(e` line
We were building a string without use it.

Author: Damn! That's silly 🤦 I assumed that line is also mutating the errors object just like the one above. Removed 😅
